### PR TITLE
Initial implement Kyuubi Chat Engine

### PR DESCRIPTION
**Pull request by pan3793**
### _Why are the changes needed?_

Introduce a brand new CHAT engine, it's supposed to support different backends, e.g. ChatGPT, 文心一言, etc.

This PR implements the following providers:

- ECHO, simply replies a welcome message.
- GPT: a.k.a ChatGPT, powered by OpenAI, which requires a API key for authentication. https://platform.openai.com/account/api-keys

Add the following configurations in `kyuubi-defaults.conf`
```
kyuubi.engine.chat.provider=[ECHO|GPT]
kyuubi.engine.chat.gpt.apiKey=<chat-gpt-api-key> 
```

Open an ECHO beeline chat engine.
```
beeline -u 'jdbc:hive2://localhost:10009/?kyuubi.engine.type=CHAT;kyuubi.engine.chat.provider=ECHO'
```

```
Connecting to jdbc:hive2://localhost:10009/
Connected to: Kyuubi Chat Engine (version 1.8.0-SNAPSHOT)
Driver: Kyuubi Project Hive JDBC Client (version 1.7.0)
Beeline version 1.7.0 by Apache Kyuubi
0: jdbc:hive2://localhost:10009/> Hello, Kyuubi!;
+----------------------------------------+
|                 reply                  |
+----------------------------------------+
| This is ChatKyuubi, nice to meet you!  |
+----------------------------------------+
1 row selected (0.397 seconds)
```

Open a ChatGPT beeline chat engine. (make sure your network can connect the open API and configure the API key)
```
beeline -u 'jdbc:hive2://localhost:10009/?kyuubi.engine.type=CHAT;kyuubi.engine.chat.provider=GPT'
```

image

### _How was this patch tested?_
- [x] Add some test cases that check the changes thoroughly including negative and positive cases if possible

- [x] Add screenshots for manual tests if appropriate

- [x] [Run test](https://kyuubi.readthedocs.io/en/master/develop_tools/testing.html#running-tests) locally before make a pull request

### CODE REVIEW

**Comment by pan3793** [comment]
Thanks @cxzl25, it works w/ ChatGPT now
image

**Review by SteNicholas — APPROVED** [review]
LGTM.

**Review by Yikf — APPROVED** [review]

File: externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+      .build(new CacheLoader[String, util.ArrayDeque[Message]] {
+        override def load(sessionId: String): util.ArrayDeque[Message] =
+          new util.ArrayDeque[Message]
+      })
+
+  override def open(sessionId: String): Unit = {
+    chatHistory.getIfPresent(sessionId)
+  }
+
+  // TODO
```
[the remark this answers is not in the document]
Comment (pan3793) [reply externals/kyuubi-chat-engine/src/main/scala/org/apache/kyuubi/engine/chat/provider/ChatGPTProvider.scala]: deleted

File: kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala [hunk 2]
```diff
@@ -2621,6 +2622,47 @@ object KyuubiConf {
...
+      .transform {
+        case "ECHO" | "echo" => "org.apache.kyuubi.engine.chat.provider.EchoProvider"
+        case "GPT" | "gpt" | "ChatGPT" => "org.apache.kyuubi.engine.chat.provider.ChatGPTProvider"
+        case other => other
+      }
+      .createWithDefault("gpt")
+
+  val ENGINE_CHAT_GPT_TOKEN: OptionalConfigEntry[String] =
+    buildConf("kyuubi.engine.chat.gpt.token")
+      .doc("The heap memory for the Chat engine")
```
[the remark this answers is not in the document]
Comment (pan3793) [reply kyuubi-common/src/main/scala/org/apache/kyuubi/config/KyuubiConf.scala]: thanks, updated

**Comment by pan3793** [comment]
Thanks all, merged to master

**Pull request closed without merging**